File: libraryBase/src/main/java/com/ashlikun/core/BaseUtils.java
package com.ashlikun.core;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.viewbinding.ViewBinding;

import com.ashlikun.core.listener.IBaseWindow;
import com.ashlikun.loadswitch.DefaultOnLoadLayoutListener;
import com.ashlikun.loadswitch.OnLoadLayoutListener;
import com.ashlikun.loadswitch.OnLoadSwitchClick;

import java.lang.reflect.AccessibleObject;
import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.HashMap;

/**
 * 作者　　: 李坤
 * 创建时间: 2019/5/24　14:43
 * 邮箱　　：496546144@qq.com
 * <p>
 * 功能介绍：工具
 */
public class BaseUtils {
    /**
     * 布局切换的布局渲染事件,必须有双参数构造方法
     */
    public static Class<? extends OnLoadLayoutListener> switchLayoutListener = null;
    private static HashMap<Class, AccessibleObject> viewBindingGetMap;

    /**
     * 本框架协成默认错误的处理,如果调用者处理了，那么这里不会调用
     * CoroutineExceptionHandler { _, t ->
     * t.printStackTrace()
     * }
     */
    public static Object coroutineExceptionHandler;


    public static void init(Class<? extends OnLoadLayoutListener> listener) {
        switchLayoutListener = listener;
    }

    /**
     * 通过反射 获取设置的全局OnLoadLayoutListener
     *
     * @param context
     * @param window
     * @return
     */
    public static OnLoadLayoutListener getSwitchLayoutListener(Context context, IBaseWindow window) {
        OnLoadLayoutListener loadLayoutListener = null;
        if (switchLayoutListener != null) {
            try {
                //获取构造函数
                Constructor con = switchLayoutListener.getConstructor(Context.class, OnLoadSwitchClick.class);
                loadLayoutListener = (OnLoadLayoutListener) con.newInstance(context, window.getOnLoadSwitchClick());
            } catch (Exception e) {
                e.printStackTrace();
            }
            if (loadLayoutListener == null) {
                try {
                    loadLayoutListener = (OnLoadLayoutListener) switchLayoutListener.newInstance();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
        if (loadLayoutListener == null) {
            loadLayoutListener = new DefaultOnLoadLayoutListener(context, window.getOnLoadSwitchClick());
        }
        return loadLayoutListener;
    }

    /**
     * 获取设置的全局OnLoadLayoutListener
     *
     * @param context
     * @return
     */
    public static OnLoadLayoutListener getSwitchLayoutListener(Context context, OnLoadSwitchClick click) {
        OnLoadLayoutListener loadLayoutListener = null;
        if (switchLayoutListener != null) {
            try {
                //获取构造函数
                Constructor con = switchLayoutListener.getConstructor(Context.class, OnLoadSwitchClick.class);
                loadLayoutListener = (OnLoadLayoutListener) con.newInstance(context, click);
            } catch (Exception e) {
                e.printStackTrace();
            }
            if (loadLayoutListener == null) {
                try {
                    loadLayoutListener = (OnLoadLayoutListener) switchLayoutListener.newInstance();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
        if (loadLayoutListener == null) {
            loadLayoutListener = new DefaultOnLoadLayoutListener(context, click);
        }
        return loadLayoutListener;
    }

    /**
     * 反射查找ViewBinding的view
     */
    public static View getViewBindingView(Object object) {
        if (object == null) {
            return null;
        }
        try {
            //检测是否有ViewBinding 库
            Class viewBindingCls = ViewBinding.class;
            if (viewBindingGetMap == null) {
                viewBindingGetMap = new HashMap<>();
            }
            Class objCls = object.getClass();
            //查找缓存
            AccessibleObject accessibleObject = null;
            if (viewBindingGetMap != null) {
                accessibleObject = viewBindingGetMap.get(objCls);
            }
            if (accessibleObject != null) {
                accessibleObject.setAccessible(true);
                if (accessibleObject instanceof Method) {
                    return ((ViewBinding) ((Method) accessibleObject).invoke(object)).getRoot();
                } else if (accessibleObject instanceof Field) {
                    return ((ViewBinding) ((Field) accessibleObject).get(object)).getRoot();
                }
            }
            Class cls = objCls;
            while (cls != null && cls != Object.class) {
                Field[] declaredFields = cls.getDeclaredFields();
                //获取字段
                for (Field f : declaredFields) {
                    if (viewBindingCls.isAssignableFrom(f.getType())) {
                        f.setAccessible(true);
                        View view = ((ViewBinding) f.get(object)).getRoot();
                        if (viewBindingGetMap == null) {
                            viewBindingGetMap = new HashMap<>();
                        }
                        viewBindingGetMap.put(objCls, f);
                        return view;
                    }
                }
                //获取方法
                Method[] declaredMethods = cls.getDeclaredMethods();
                for (Method m : declaredMethods) {
                    if (viewBindingCls.isAssignableFrom(m.getReturnType())) {
                        m.setAccessible(true);
                        View view = ((ViewBinding) m.invoke(object)).getRoot();
                        if (viewBindingGetMap == null) {
                            viewBindingGetMap = new HashMap<>();
                        }
                        viewBindingGetMap.put(objCls, m);
                        return view;
                    }
                }
                //获取父类的
                cls = cls.getSuperclass();
            }
        } catch (Exception e) {
            return null;
        } catch (NoClassDefFoundError e) {
            return null;
        }
        return null;
    }

    /**
     * 获取1个参数的静态方法
     */
    public static View getViewToViewBindingClass(Class cls, LayoutInflater layoutInflater) {
        Object viewBindingToClass = getViewBindingToClass(cls, layoutInflater);
        try {
            if (viewBindingToClass != null) {
                if (viewBindingToClass instanceof ViewBinding) {
                    ViewBinding bb = (ViewBinding) viewBindingToClass;
                    return bb.getRoot();
                }
            }
        } catch (NoClassDefFoundError e) {
            return null;
        }
        return null;
    }

    /**
     * 获取1个参数的静态方法
     *
     * @return ViewBinding
     */
    public static Object getViewBindingToClass(Class cls, LayoutInflater layoutInflater) {
        if (cls == null || layoutInflater == null) {
            return null;
        }
        boolean isCache = false;
        //从缓存获取
        try {
            Method inflate = null;
            if (viewBindingGetMap != null) {
                AccessibleObject aa = viewBindingGetMap.get(cls);
                if (aa instanceof Method) {
                    isCache = true;
                    inflate = (Method) aa;
                }
            }
            if (inflate != null) {
                inflate = cls.getDeclaredMethod("inflate", LayoutInflater.class);
            }
            //这里循环全部方法是为了混淆的时候无影响
            if (inflate == null) {
                Method[] declaredMethods = cls.getDeclaredMethods();
                for (Method declaredMethod : declaredMethods) {
                    int modifiers = declaredMethod.getModifiers();
                    if (Modifier.isStatic(modifiers)) {
                        Class<?>[] parameterTypes = declaredMethod.getParameterTypes();
                        if (parameterTypes != null && parameterTypes.length == 3) {
                            if (LayoutInflater.class.isAssignableFrom(parameterTypes[0]) &&
                                    ViewGroup.class.isAssignableFrom(parameterTypes[1]) &&
                                    boolean.class.isAssignableFrom(parameterTypes[2])) {
                                inflate = declaredMethod;
                                break;
                            }
                        }
                    }
                }
            }
            if (inflate != null) {
                //添加到缓存
                if (isCache) {
                    if (viewBindingGetMap == null) {
                        viewBindingGetMap = new HashMap<>();
                    }
                    viewBindingGetMap.put(cls, inflate);
                }
                return inflate.invoke(null, layoutInflater);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

}

File: libraryBase/src/main/java/com/ashlikun/core/BaseUtils.kt
package com.ashlikun.core

import android.app.Activity
import android.content.Context
import android.content.res.Configuration
import android.content.res.Resources
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import androidx.viewbinding.ViewBinding
import com.ashlikun.core.listener.IBaseWindow
import com.ashlikun.loadswitch.DefaultOnLoadLayoutListener
import com.ashlikun.loadswitch.OnLoadLayoutListener
import com.ashlikun.loadswitch.OnLoadSwitchClick
import com.ashlikun.utils.other.coroutines.defaultCoroutineExceptionHandler
import kotlinx.coroutines.CoroutineExceptionHandler
import java.lang.reflect.*
import kotlin.coroutines.CoroutineContext

/**
 * @author　　: 李坤
 * 创建时间: 2021/12/19 19:53
 * 邮箱　　：496546144@qq.com
 *
 * 功能介绍：工具
 */
object BaseUtils {
    /**
     * Activity onCreate创建之前
     */
    internal var activityPreCreated: MutableList<((activity: Activity, savedInstanceState: Bundle?) -> Unit)>? = null
    val onActivityPreCreated
        get() = if (activityPreCreated == null) {
            activityPreCreated = mutableListOf()
            activityPreCreated!!
        } else activityPreCreated!!

    /**
     * Activity onAttachBaseContext
     */
    internal var attachBaseContext: MutableList<((context: Context) -> Context)>? = null
    val onAttachBaseContext
        get() = if (attachBaseContext == null) {
            attachBaseContext = mutableListOf()
            attachBaseContext!!
        } else attachBaseContext!!

    /**
     * Activity applyOverrideConfiguration
     */
    internal var applyOverrideConfiguration: MutableList<((configuration: Configuration) -> Configuration)>? = null
    val onApplyOverrideConfiguration
        get() = if (applyOverrideConfiguration == null) {
            applyOverrideConfiguration = mutableListOf()
            applyOverrideConfiguration!!
        } else applyOverrideConfiguration!!

    /**
     * Activity onConfigurationChanged
     */
    internal var configurationChanged: MutableList<((newConfig: Configuration) -> Configuration)>? = null
    val onConfigurationChanged
        get() = if (configurationChanged == null) {
            configurationChanged = mutableListOf()
            configurationChanged!!
        } else configurationChanged!!

    /**
     * 当调用Activity的getResources将被调用，便于hook,只调用一次，内部会缓存
     */
    internal var activityGetResources: MutableList<((result: Resources) -> Resources)>? = null
    val onActivityGetResources
        get() = if (activityGetResources == null) {
            activityGetResources = mutableListOf()
            activityGetResources!!
        } else activityGetResources!!

    /**
     * 全局当Activity回调 onUserInteraction 方法时候
     */
    internal var userInteraction: MutableList<((activity: Activity) -> Unit)>? = null
    val onUserInteraction
        get() = if (userInteraction == null) {
            userInteraction = mutableListOf()
            userInteraction!!
        } else userInteraction!!

    /**
     * 全局当Activity回调 dispatchTouchEvent 方法时候
     */
    internal var dispatchTouchEvent: MutableList<((activity: Activity, ev: MotionEvent) -> Boolean)>? = null
    val onDispatchTouchEvent
        get() = if (dispatchTouchEvent == null) {
            dispatchTouchEvent = mutableListOf()
            dispatchTouchEvent!!
        } else dispatchTouchEvent!!

    /**
     * 布局切换的布局渲染事件,必须有双参数构造方法
     */
    var switchLayoutListener: ((context: Context, click: OnLoadSwitchClick?) -> OnLoadLayoutListener?)? = null

    /**
     * ViewBinding反射的缓存
     */
    val viewBindingGetMap = mutableMapOf<Class<*>, AccessibleObject>()

    /**
     * 本框架协成默认错误的处理,如果调用者处理了，那么这里不会调用
     * CoroutineExceptionHandler { _, t ->
     * t.printStackTrace()
     * }
     * 如果不设置就会调用工具类的defaultCoroutineExceptionHandler
     */
    var coroutineExceptionHandler: CoroutineExceptionHandler? = null
        get() = field ?: defaultCoroutineExceptionHandler


    /**
     *  获取设置的全局OnLoadLayoutListener
     */
    fun createSwitchLayoutListener(window: IBaseWindow) =
        createSwitchLayoutListener(window.requireContext, window.onLoadSwitchClick)

    /**
     * 获取设置的全局OnLoadLayoutListener
     */
    fun createSwitchLayoutListener(context: Context, click: OnLoadSwitchClick?) =
        switchLayoutListener?.invoke(context, click) ?: DefaultOnLoadLayoutListener(context, click)

    /**
     * 反射查找ViewBinding的view
     */
    fun getViewBindingView(obj: Any): View? {
        try {
            //检测是否有ViewBinding 库
            val viewBindingCls: Class<out ViewBinding> = ViewBinding::class.java
            val objCls: Class<*> = obj.javaClass
            //查找缓存
            var accessibleObject = viewBindingGetMap[objCls]
            if (accessibleObject != null) {
                accessibleObject.isAccessible = true
                if (accessibleObject is Method) {
                    return (accessibleObject.invoke(obj) as ViewBinding).root
                } else if (accessibleObject is Field) {
                    return (accessibleObject[obj] as ViewBinding).root
                }
            }
            var cls: Class<*>? = objCls
            while (cls != null && cls != Any::class.java) {
                val declaredFields = cls.declaredFields
                //获取字段
                for (f in declaredFields) {
                    if (viewBindingCls.isAssignableFrom(f.type)) {
                        f.isAccessible = true
                        val view = (f[obj] as ViewBinding).root
                        viewBindingGetMap[objCls] = f
                        return view
                    }
                }
                //获取方法
                val declaredMethods = cls.declaredMethods
                for (m in declaredMethods) {
                    if (viewBindingCls.isAssignableFrom(m.returnType)) {
                        m.isAccessible = true
                        val view = (m.invoke(obj) as ViewBinding).root
                        viewBindingGetMap[objCls] = m
                        return view
                    }
                }
                //获取父类的
                cls = cls.superclass
            }
        } catch (e: Exception) {
            return null
        } catch (e: NoClassDefFoundError) {
            return null
        }
        return null
    }

    /**
     * 获取1个参数的静态方法
     */
    @JvmStatic
    fun getViewToViewBindingClass(cls: Class<*>?, layoutInflater: LayoutInflater?): View? {
        val viewBindingToClass = getViewBindingToClass(cls, layoutInflater)
        try {
            if (viewBindingToClass != null) {
                if (viewBindingToClass is ViewBinding) {
                    return viewBindingToClass.root
                }
            }
        } catch (e: NoClassDefFoundError) {
            return null
        }
        return null
    }

    /**
     * 获取1个参数的静态方法
     *
     * @return ViewBinding
     */
    fun getViewBindingToClass(cls: Class<*>?, layoutInflater: LayoutInflater?): Any? {
        if (cls == null || layoutInflater == null) {
            return null
        }
        var isCache = false
        //从缓存获取
        try {
            var inflate: Method? = null
            val cacheValue = viewBindingGetMap[cls]
            if (cacheValue is Method) {
                isCache = true
                inflate = cacheValue
            }
            if (inflate != null) {
                inflate = cls.getDeclaredMethod("inflate", LayoutInflater::class.java)
            }
            //这里循环全部方法是为了混淆的时候无影响
            if (inflate == null) {
                val declaredMethods = cls.declaredMethods
                for (declaredMethod in declaredMethods) {
                    val modifiers = declaredMethod.modifiers
                    if (Modifier.isStatic(modifiers)) {
                        val parameterTypes = declaredMethod.parameterTypes
                        if (parameterTypes != null && parameterTypes.size == 3) {
                            if (LayoutInflater::class.java.isAssignableFrom(parameterTypes[0]) &&
                                ViewGroup::class.java.isAssignableFrom(parameterTypes[1]) &&
                                Boolean::class.javaPrimitiveType!!.isAssignableFrom(parameterTypes[2])
                            ) {
                                inflate = declaredMethod
                                break
                            }
                        }
                    }
                }
            }
            if (inflate != null) {
                //添加到缓存
                if (isCache) {
                    viewBindingGetMap[cls] = inflate
                }
                return inflate.invoke(null, layoutInflater)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }
}